src/OpenTelemetry.Contrib.Instrumentation.Remoting/Implementation/TelemetryDynamicSink.cs

```diff
@@ -0,0 +1,232 @@
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+// </copyright>
+
+using System;
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Linq;
+using System.Runtime.Remoting.Contexts;
+using System.Runtime.Remoting.Messaging;
+using OpenTelemetry.Context.Propagation;
+using OpenTelemetry.Trace;
+
+namespace OpenTelemetry.Contrib.Instrumentation.Remoting.Implementation
+{
+    /// <summary>
+    /// A dynamic remoting sink that intercepts all calls leaving or entering an AppDomain, see
+    /// <see cref="RemotingInstrumentation"/> constructor: RegisterDynamicProperty
+    /// code.
+    /// </summary>
+    /// <remarks>
+    /// See https://docs.microsoft.com/en-us/previous-versions/dotnet/netframework-4.0/tdzwhfy3(v=vs.100) for
+    /// information about remoting sinks and https://docs.microsoft.com/en-us/dotnet/api/system.runtime.remoting.contexts.context.registerdynamicproperty?view=netframework-4.8
+    /// for RegisterDynamicProperty behavior.
+    /// </remarks>
+    internal class TelemetryDynamicSink : IDynamicMessageSink
+    {
+        internal const string AttributeRpcSystem = "rpc.system";
+        internal const string AttributeRpcService = "rpc.service";
+        internal const string AttributeRpcMethod = "rpc.method";
+
+        // Uri like "tcp://localhost:1234/HelloServer.rem"
+        // TODO: semantic conventions don't have an attribute for a full uri of an RPC endpoint, but seems useful?
+        internal const string AttributeRpcRemotingUri = "rpc.netframework_remoting.uri";
+
+        internal const string ActivitySourceName = "OpenTelemetry.Remoting";
+        private const string ActivityOutName = ActivitySourceName + ".RequestOut";
+        private const string ActivityInName = ActivitySourceName + ".RequestIn";
+
+        private static readonly Version Version = typeof(TelemetryDynamicSink).Assembly.GetName().Version;
+        private static readonly ActivitySource RemotingActivitySource = new ActivitySource(ActivitySourceName, Version.ToString());
+
+        private static readonly ConcurrentDictionary<string, string> ServiceNameCache = new ConcurrentDictionary<string, string>();
+
+        private readonly RemotingInstrumentationOptions options;
+
+        public TelemetryDynamicSink(RemotingInstrumentationOptions options)
+        {
+            this.options = options;
+        }
+
+        public void ProcessMessageStart(IMessage reqMsg, bool bCliSide, bool bAsync)
```

Comment:
what is the impact if this method throws an exception? Does it make sense to wrap the entire method inside try..catch..log blocks?

Reply:
Thanks - that would have been an appdomain unhandled exception! Added try-catch-log for both start and finish.